proposals/0272-sbpf-encoding-efficiency.md

```diff
@@ -0,0 +1,162 @@
+---
+simd: '0272'
+title: SBPF Encoding Efficiency
+authors:
+  - Alexander Meißner
+  - Lucas Steuernagel
+category: Standard
+type: Core
+status: Idea
+created: 2025-04-01
+feature: TBD
+extends: SIMD-0161
+---
+
+## Summary
+
+Improve encoding efficiency in SBPF-v5.
+
+## Motivation
+
+SBPF inherited the 64 bit intruction layout from BPF including its space
+inefficiency. For example the 16 bit displacement is only used for memory
+access and conditional branch instructions and the 32 bit immediate value is
+only used in instructions with an immediate operand. Yet, all other
+instructions still need to encode them.
+
+Additionally, there are only 10 general purpose registers which leads to a lot
+of stack spilling and thus additional instructions and larger executable files.
+Unlike typical CISC instruction sets like x86, which do have memory indirect
+operands to compensate for their low number of general purpose registers, BPF
+and thus in turn SBPF has no such thing.
+
+Reducing the instruction frame size from 64 bit to 32 bit and increasing the
+addressable registers from 11 to 32 should dramatically improve the encoding
+efficiency and allow for bigger (meaning more complex) on-chain programs and
+lower rent excemption funds being required for programs at the same complexity.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+### Register Layouts
+
+The current register layout from SBPF-v1 is:
+
+|         name | kind            | Solana ABI
+|-------------:|:----------------|:----------
+| `r0`         | GPR             | Return value
+| `r1` to `r5` | GPR             | Argument registers
+| `r6` to `r9` | GPR             | Call-preserved
+| `r10`        | Frame pointer   | System register
+| `pc`         | Program counter | Hidden register
+
+The register layout from SBPF-v5 on is:
+
+|           name | kind            | Solana ABI
+|---------------:|:----------------|:----------
+| `r0`           | GPR             | Return value
+| `r1` to `r5`   | GPR             | Argument registers
```

Comment:
I've seen many function calls use more than five registers to store arguments, so allocating more registers here (six or seven) would be beneficial to avoid stack spills.